Hacktoberfest [JENKINS-69731] Update shared-libraries.adoc

jimklimov opened the pull request:
Document limited run-time evaluated version for `@Library` annotation.

See also:
* https://github.com/jenkinsci/pipeline-groovy-lib-plugin/pull/19
* https://issues.jenkins.io/browse/JENKINS-69731

jimklimov committed:
Update shared-libraries.adoc

Document limited run-time evaluated version for `@Library` annotation.

https://github.com/jenkinsci/pipeline-groovy-lib-plugin/pull/19
https://issues.jenkins.io/browse/JENKINS-69731

jimklimov committed:
Update shared-libraries.adoc

Fix pass-through markup for website generator

MarkEWaite commented on content/doc/book/pipeline/shared-libraries.adoc:
```diff
@@ -194,6 +194,54 @@ A link to the example repository used is also provided in the link:https://youtu
 
+=== Statically loading evaluated versions of libraries
+
+As of currently unreleased version of the _Pipeline: Shared Groovy Libraries_ plugin,
+with features proposed in
+link:https://github.com/jenkinsci/pipeline-groovy-lib-plugin/pull/19[PR #19]
+(link:https://issues.jenkins.io/browse/JENKINS-69731[issue JENKINS-69731])
```
Thanks for phrasing it that way to make it clear to reviewers like me that the feature is not yet released.  Very nice technique!
```suggestion
```

jimklimov replied:
I suppose you did not mean removing these lines, just yet? :)

(If this PR is indeed only merged after the feature release, this should get rephrased similar to `library` step description)

MarkEWaite commented on content/doc/book/pipeline/shared-libraries.adoc:
```diff
@@ -194,6 +194,54 @@ A link to the example repository used is also provided in the link:https://youtu
+@Library('my-shared-library@${BRANCH_NAME}') _
+----
+
+The plugin would try several methods to determine the branch name to use, check if
+such "version" is accepted by SCM source of the library, and fall back to default
+version if not.
```
Simpler verbs and a few minor changes for your consideration
```suggestion
The plugin tries several methods to determine the branch name to use, checks if
the "version" is accepted by the SCM source of the library, and falls back to the default
version if not.
```

MarkEWaite commented on content/doc/book/pipeline/shared-libraries.adoc:
```diff
@@ -194,6 +194,54 @@ A link to the example repository used is also provided in the link:https://youtu
+
+Another addressed use-case regards loading of some version of library specified by
+an environment variable value defined when the job starts (e.g. coming from Global
+Jenkins configuration, from build agent configuration, or injected by some plugins):
```
```suggestion
Jenkins configuration, from build agent configuration, or injected by a plugin):
```

MarkEWaite commented on content/doc/book/pipeline/shared-libraries.adoc:
```diff
@@ -194,6 +194,54 @@ A link to the example repository used is also provided in the link:https://youtu
+* Pipelines backed by a single SCM source with a "Branch Specifier" like `*/main`;
+* "In-line" Pipelines not backed by an SCM (just fall back to default version).
+
+Another addressed use-case regards loading of some version of library specified by
```
Possibly simpler phrasing?
```suggestion
The plugin can also load a library version specified by
```

jimklimov replied:
Not sure about this change (alone) since an earlier paragraph stressed that "one use-case is..." so here "another"...

MarkEWaite reviewed (approved):
Looks good to me, with some optional comments for your consideration.

This should not be merged until the pull request has been merged into the plugin and the plugin has been released.

jimklimov committed:
Update content/doc/book/pipeline/shared-libraries.adoc

Better wording from PR review

Co-authored-by: Mark Waite <mark.earl.waite@gmail.com>

jimklimov committed:
Update content/doc/book/pipeline/shared-libraries.adoc

Better wording from PR review

Co-authored-by: Mark Waite <mark.earl.waite@gmail.com>

jmMeessen commented:
Hello @jimklimov ,

I am Jenkins Developer Relations and promoting Hacktoberfest@Jenkins. I am collecting stories of this year's participants to share with the community. 

First of all, thank you very much for contributing to Jenkins during this Hacktoberfest. It is much appreciated.

This is why I am reaching out to you to better get to know you as a community member and a contributor. Would you mind to answer the following questions?

- Who are you? Where are you located? What do you do for a living?
- Is it the first time you participate to Hacktoberfest?
- What are your motivations?
- Did you already contribute to Open Source before?
- Why is Open Source important to you (as you are contributing)?
- What did you contribute to during this hacktoberfest?
- What do you experience as most difficult when contributing to open source?
- What would you say to others that are hesitating to contribute (because of lack of skills/fear/...)

You can reach me privately via jean-marc@meessen-web.org. 

If it is easier for you I could setup a call during one of your evenings (I am located in Europe) and have a short chat together. I would love to hear from you so that I have a better grasp of the multiple facets of our Community.

If you answer in this comment, by mail (or if we setup a call), would you be OK if I share your story (anonymously or by mentioning your name) with the Community? It would be either during a Hacktoberfest related livestream or a blog post. I strongly believe that sharing real human stories would motivate people to make the step and join us during this month or later.

/- Jmm

Brussels, Belgium
jean-marc@meessen-web.org

jimklimov commented:
Cheers @jmMeessen - I suppose we've met on some of the FOSDEMs so you might have some answers already, but I understand this is a neat questionnaire anyway :) Feel free to arrange and share:

> This is why I am reaching out to you to better get to know you as a community member and a contributor. Would you mind to answer the following questions?

Sure, but in a bit of a different order for a better narrative :)

>     * Who are you? Where are you located? What do you do for a living?

I am Jim Klimov, and my studies and further career took me through academia (minoring in "Open Information Systems", no less) through systems administration in a college campus, then in a company spun off from an university lab to deal with various enterprise document workflow and messaging systems, raising to CTO of that small collective, later moving to Czech Republic and working in a couple of corporate settings.

During my path I discovered a few things about myself and the world:
* that no tool made by others meets all of our needs perfectly - so there indeed is a need to keep forging them, and preferably get someone else to maintain the fixes long-term (hello Open Source and contributions);
* that I like "to forge the tools of my trade" (mostly by scripting);
* that even though I am not too proficient as a programmer, I can get things done in any language (people did say my Java looked like shell script... but hey, it works);
* and that I like to help "real programmers" get their job done well so that a project can ultimately reach its users.

Currently I am a "DevOps architect", working closely with product architects and QA leaders, helping a company design and implement how the product code they write and run on several platforms would become a well-tested, upgradable and reliable deliverable for end-users (as on-premise or managed-cloud service), as they crush a legacy monolithic application into a swarm of services.

>     * Did you already contribute to Open Source before?

Yes, starting last millennium (in the university network administration), posting some patches to Samba and exchanging ideas on various Linux mailing lists. Later on I collaborated on several milters (sendmail filters) for anti-spam, SDIG (Switch Digger) and NUT to monitor our infra hardware, OpenVPN, ClamAV, sort of maintained a campus baseline Linux distro recipes built with optimization for each ARCH of very different computers posing as servers, etc. On the professional side we worked with Sun Microsystems projects a lot (OS, terminal services, VMs for workspaces, messaging suite, Java app servers, ZFS and NFS based file services being a tip of the iceberg), so learned a lot from their engineering blogs and I frequented the forums and mailing lists to share back tips and tricks that I discovered in their products - in a way, this was Open Source knowledge about a lot of proprietary products.

Later on many of these Sun projects got opened up and often forked as OpenSolaris (now illumos and many distros based on it), Hudson (now Jenkins), Glassfish (now Payara), OpenLDAP and OpenIDM (picked up by ForgeRock for a while), NetBeans, VirtualBox, etc. This grand opening in early 2000's let us systems engineers see the code running in production builds (or its close relative) to understand whether something not working is our mistake (e.g. misunderstanding of configuration) or we found an issue or corner case in their product - to ultimately work around it for the current build while proposing a bug report or fix ideas to the upstream. This went on after the "Sun-set" as the projects moved on, forked often by same teams that created them and communities that used them - who could now become collaborators and co-authors.

One thing led to another, professionally and enthusiastically, and I've participated a lot in packaging maintenance for Solaris, OpenIndiana and OmniOS (all with different recipe techs and approaches), Debian-ish and RPM-ish systems, and later got introduced to OBS (OpenSUSE Build System) to automate much of that for Linux at least - and surely posted PRs to scratch some itches there too. I have maintained Jenkins servers for the past decade in several companies, introducing Pipelines as they appeared (so redoing a lot as the idea grew and early-release APIs changed), culminating in contributions to libzfs4j (so it accepts more native libzfs APIs - which are internal and incompatible over the years - and Jenkins does not crash upon startup) on one of the FOSDEM meet-ups, and ending up with a take-over (and releasing the reins) of some Jenkins plugins where functionality lacked for us, and current maintainers could not merge the fixes in a timely fashion :)

>     * Why is Open Source important to you (as you are contributing)?

 
>     * Is it the first time you participate to Hacktoberfest?
> 
>     * What are your motivations?
> 
>     * What did you contribute to during this hacktoberfest?
> 
>     * What do you experience as most difficult when contributing to open source?
> 
>     * What would you say to others that are hesitating to contribute (because of lack of skills/fear/...)
> 
> 
> You can reach me privately via [jean-marc@meessen-web.org](mailto:jean-marc@meessen-web.org).
> 
> If it is easier for you I could setup a call during one of your evenings (I am located in Europe) and have a short chat together. I would love to hear from you so that I have a better grasp of the multiple facets of our Community.
> 
> If you answer in this comment, by mail (or if we setup a call), would you be OK if I share your story (anonymously or by mentioning your name) with the Community? It would be either during a Hacktoberfest related livestream or a blog post. I strongly believe that sharing real human stories would motivate people to make the step and join us during this month or later.
> 
> /- Jmm
> 
> Brussels, Belgium [jean-marc@meessen-web.org](mailto:jean-marc@meessen-web.org)

jimklimov commented:
@jmMeessen : sorry the post above took several attempts, so if you see it incomplete via GitHub mail notifications - please revise the final version in the web.

jimklimov committed:
Update shared-libraries.adoc

Update the vision on `@Library('my-shared-library@${env.CUSTOM_VAR_NAME}')` - that "build agents" probably have no say in the envvar value, yet the feature is not useless :)

jmMeessen commented:
Hey Jim (@jimklimov )

Thank you very much for having taken the time to write this extensive reply on my questions. Your answers are very rich. 

I don't remember if we already met at a FOSDEM. But I'll be present at this year's edition. If you can make it to Brussels, it would be my pleasure to continue this discussion "face to face".  I'll be wearing my captain's hat (reference to a european commercial) or, maybe, my Santa hat (better known reference). I will be easily recognizable 😄 

Thank again for your contribution and for sharing your thoughts.

jimklimov commented:
@jmMeessen : Slightly updated my post here and there (just in case you copied it for publication etc.) :)

jimklimov committed:
Enhance documentation on library loading and security [JENKINS-69731]

* Added security considerations and usage guidelines for dynamically resolved library revisions in Jenkins pipelines.
* Noted details on PR support (resolution via source/target branch names)
* For dynamic library revision support, a way was added to restrict allowed revision patterns - recommended in docs now
